src/main/java/com/limechain/runtime/allocator/Header.java

```diff
@@ -0,0 +1,81 @@
+package com.limechain.runtime.allocator;
+
+import lombok.Getter;
+import org.wasmer.Memory;
+
+/**
+ * Allocation header preceding a memory block.
+ * <br>
+ * Can be one of two types:
+ * <ul>
+ *     <li>Free: containing a pointer to the next free block and signifying that the block after it is free.</li>
+ *     <li>
+ *         Occupied: containing the {@link Order order} of the block following it and signifying that it is occupied.
+ *     </li>
+ * </ul>
+ * The header is written in memory as 64 bits: the most significant ones denoting its type
+ * and the least significant - the header data (next free block or order).
+ */
+@Getter
+public class Header {
+    private static final long OCCUPIED_HEADER_MASK = 0x00000001_00000000L;
+    private static final long FREE_HEADER_MASK = 0x00000000_00000000L;
+
+    private final boolean occupied;
+
+    private Integer nextFreeHeaderPointer;
+    private Integer order;
+
+    private Header(boolean occupied) {
+        this.occupied = occupied;
+    }
```

Comment:
this could be a `@RequiredArgsConstructor` with accessibility argument set to private